deduce.go

```diff
@@ -519,76 +524,166 @@ func (m vcsExtensionDeducer) deduceSource(path string, u *url.URL) (maybeSource,
 	}
 }
 
-type stringFuture func() (string, error)
-type sourceFuture func() (source, string, error)
-type partialSourceFuture func(string, ProjectAnalyzer) sourceFuture
+type deductionCoordinator struct {
+	callMgr  *callManager
+	rootxt   *radix.Tree
+	deducext *deducerTrie
+	action   chan func()
+}
+
+func newDeductionCoordinator(cm *callManager) *deductionCoordinator {
+	dc := &deductionCoordinator{
+		callMgr:  cm,
+		rootxt:   radix.New(),
+		deducext: pathDeducerTrie(),
+		action:   make(chan func()),
+	}
+
+	// Start listener loop
+	go func() {
+		for {
+			select {
+			case <-dc.callMgr.getLifetimeContext().Done():
+				close(dc.action)
+				return
+			case action := <-dc.action:
+				action()
+			}
+		}
+	}()
 
-type deductionFuture struct {
-	// rslow indicates that the root future may be a slow call (that it has to
-	// hit the network for some reason)
-	rslow bool
-	root  stringFuture
-	psf   partialSourceFuture
+	return dc
 }
 
-// deduceFromPath takes an import path and attempts to deduce various
+// deduceRootPath takes an import path and attempts to deduce various
 // metadata about it - what type of source should handle it, and where its
 // "root" is (for vcs repositories, the repository root).
 //
-// The results are wrapped in futures, as most of these operations require at
-// least some network activity to complete. For the first return value, network
-// activity will be triggered when the future is called. For the second,
-// network activity is triggered only when calling the sourceFuture returned
-// from the partialSourceFuture.
-func (sm *SourceMgr) deduceFromPath(path string) (deductionFuture, error) {
-	opath := path
-	u, path, err := normalizeURI(path)
-	if err != nil {
-		return deductionFuture{}, err
-	}
+// If no errors are encountered, the returned pathDeduction will contain both
+// the root path and a list of maybeSources, which can be subsequently used to
+// create a handler that will manage the particular source.
+func (dc *deductionCoordinator) deduceRootPath(ctx context.Context, path string) (pathDeduction, error) {
+	if dc.callMgr.getLifetimeContext().Err() != nil {
+		return pathDeduction{}, errors.New("deductionCoordinator has been terminated")
+	}
+
+	retchan, errchan := make(chan pathDeduction), make(chan error)
+	dc.action <- func() {
+		hmdDeduce := func(hmd *httpMetadataDeducer) {
+			pd, err := hmd.deduce(ctx, path)
+			if err != nil {
+				errchan <- err
+			} else {
+				retchan <- pd
+			}
+		}
+
+		// First, check the rootxt to see if there's a prefix match - if so, we
+		// can return that and move on.
+		if prefix, data, has := dc.rootxt.LongestPrefix(path); has && isPathPrefixOrEqual(prefix, path) {
+			switch d := data.(type) {
```

Comment:
yes, def worth doing. i plan to do a docs pass once this stabilizes a bit more (though pretty close now); i tend to avoid doing too much of that when i'm mid-refactor, as i end up chucking the docs.